src/main/java/org/broadinstitute/hellbender/utils/nio/SeekableByteChannelPrefetcher.java

```diff
@@ -227,17 +248,38 @@ public int read(ByteBuffer dst) throws IOException {
             src = fetch(position);
             msWaitingForData += waitingForData.elapsed(TimeUnit.MILLISECONDS);
         } catch (InterruptedException e) {
+            // Restore interrupted status
+            Thread.currentThread().interrupt();
             return 0;
         } catch (ExecutionException e) {
             throw new RuntimeException(e);
         }
+        if (null == src) {
+            // the caller is asking for a block past EOF
+            nbReadsPastEnd++;
+            betweenCallsToRead.reset();
+            betweenCallsToRead.start();
+            return -1; // EOF
+        }
         Stopwatch copyingData = Stopwatch.createStarted();
         int bytesToCopy = dst.remaining();
         byte[] array = src.array();
         // src.position is how far we've written into the array
         long blockIndex = position / bufSize;
         int offset = (int)(position - (blockIndex * bufSize));
+        // src |==============---------------------|
+        //     :<---src.pos-->------src.limit----->:
+        // |---:--position->
+        //     :<--offset-->
+        //     ^ blockIndex*bufSize
         int availableToCopy = src.position() - offset;
+        if (availableToCopy < 0) {
+            // the caller is asking to read past the end of the file
+            nbReadsPastEnd++;
```

Comment:
An even better way to factor this out is with a try-finally since we need betweenCallsToRead to be reset when the method returns, regardless of how that happens. I'll put that in.